core/src/main/scala/flatgraph/DiffGraphApplier.scala

```diff
@@ -223,43 +237,60 @@ private[flatgraph] class DiffGraphApplier(graph: Graph, diff: DiffGraphBuilder,
     val ndiff = splitUpdate()
     diff.buffer = null
 
-    // set edge properties
-    for {
-      nodeKind  <- graph.schema.nodeKinds
-      edgeKind  <- graph.schema.edgeKinds
-      direction <- Edge.Direction.values
-    } setEdgeProperty(nodeKind, direction, edgeKind)
-
-    // remove edges
-    for {
-      nodeKind  <- graph.schema.nodeKinds
-      edgeKind  <- graph.schema.edgeKinds
-      direction <- Edge.Direction.values
-    } deleteEdges(nodeKind, direction, edgeKind)
-
-    // add nodes
-    for (nodeKind <- graph.schema.nodeKinds)
-      addNodes(nodeKind)
-
-    // delete nodes
-    if (delNodes.nonEmpty) {
-      deleteNodes()
-    }
-
-    // add edges
-    for {
-      nodeKind  <- graph.schema.nodeKinds
-      edgeKind  <- graph.schema.edgeKinds
-      direction <- Direction.values
-    } addEdges(nodeKind, direction, edgeKind)
+    try {
+      // set edge properties
+      for {
+        nodeKind  <- graph.schema.nodeKinds
+        edgeKind  <- graph.schema.edgeKinds
+        direction <- Edge.Direction.values
+      } setEdgeProperty(nodeKind, direction, edgeKind)
+
+      jobQueue.foreach(_.get())
+      jobQueue.clear()
+
+      // remove edges
+      for {
+        nodeKind  <- graph.schema.nodeKinds
+        edgeKind  <- graph.schema.edgeKinds
+        direction <- Edge.Direction.values
+      } deleteEdges(nodeKind, direction, edgeKind)
+
+      jobQueue.foreach(_.get())
+      jobQueue.clear()
+
+      // add nodes. If necessary, we can start this job earlier in a separate queue; then we must add some extra structures for node sizes at various stages
+      for { nodeKind <- graph.schema.nodeKinds } addNodes(nodeKind)
+
+      jobQueue.foreach(_.get())
+      jobQueue.clear()
+
+      // delete nodes
+      // this is harder to parallelize, only do if needed!
+      if (delNodes.nonEmpty) {
+        deleteNodes()
+      }
 
-    // set node properties
-    for (nodeKind <- graph.schema.nodeKinds) {
-      for (propertyKind <- graph.schema.propertyKinds) {
-        setNodeProperties(nodeKind, propertyKind)
+      // add edges
+      for {
+        nodeKind  <- graph.schema.nodeKinds
+        edgeKind  <- graph.schema.edgeKinds
+        direction <- Direction.values
+      } addEdges(nodeKind, direction, edgeKind)
+
+      // set node properties. This doesn't need to wait for add edges to be complete!
+      for (nodeKind <- graph.schema.nodeKinds) {
+        for { propertyKind <- graph.schema.propertyKinds } setNodeProperties(nodeKind, propertyKind)
+        if (newNodeUsers.decrementAndGet(nodeKind) == -1) {
+          // whoever reaches newNodeUsers(nodeKind) == -1 first is permitted to free it.
+          newNodes(nodeKind) = null
+        }
       }
-      // we can now clear the newnodes
-      newNodes(nodeKind) = null
+
+      jobQueue.foreach(_.get())
+      jobQueue.clear()
+    } catch {
+      case ex: java.util.concurrent.ExecutionException =>
+        throw ex.getCause()
```

Comment:
this will just throw away the root exception, right? why?

Reply:
In order to reproduce the old behavior (which is also tested for!). The whole exectutor-stuff wraps exceptions, and people who try to catch e.g. SchemaViolation exceptions presumably want their exceptions unwrapped.

Reply:
that makes sense - can you copy paste your response into the codebase please?